File: android/VoiceRecorder-Calculator/app/src/main/java/com/example/waveform/MainActivity.java
package com.example.waveform;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity {
    private static final int PERMISSION_REQUEST_CODE = 100;
    private boolean pendingOpenRecorder = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        View cardRecorder = findViewById(R.id.cardRecorder);
        cardRecorder.setOnClickListener(v -> {
            pendingOpenRecorder = true;
            if (checkAllPermissions()) {
                openRecorder();
            }
        });

        View cardCalculator = findViewById(R.id.cardCalculator);
        cardCalculator.setOnClickListener(v -> {
            Intent intent = new Intent(this, CalculatorActivity.class);
            startActivity(intent);
        });
    }

    private void openRecorder() {
        RecorderBottomSheetFragment bottomSheet = new RecorderBottomSheetFragment();
        bottomSheet.show(getSupportFragmentManager(), "RecorderSheet");
        pendingOpenRecorder = false;
    }

    private boolean checkAllPermissions() {
        // 检查录音权限（必需）
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO}, PERMISSION_REQUEST_CODE);
            return false;
        }

        // 检查存储权限（Android 10以下需要，Android 10+使用分区存储）
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.Q) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERMISSION_REQUEST_CODE);
                return false;
            }
        }

        return true;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == PERMISSION_REQUEST_CODE) {
            boolean audioGranted = false;
            boolean storageGranted = true; // 默认为true，因为Android 10+不需要

            for (int i = 0; i < permissions.length; i++) {
                String permission = permissions[i];
                if (permission.equals(Manifest.permission.RECORD_AUDIO)) {
                    audioGranted = grantResults[i] == PackageManager.PERMISSION_GRANTED;
                } else if (permission.equals(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                    storageGranted = grantResults[i] == PackageManager.PERMISSION_GRANTED;
                }
            }

            if (audioGranted && storageGranted) {
                Toast.makeText(this, "权限已授予", Toast.LENGTH_SHORT).show();
                if (pendingOpenRecorder) {
                    openRecorder();
                }
            } else {
                String missingPermission = !audioGranted ? "录音权限" : "存储权限";
                Toast.makeText(this, "需要" + missingPermission + "才能使用功能", Toast.LENGTH_LONG).show();
                pendingOpenRecorder = false;
            }
        }
    }
}

File: android/VoiceWave-Android/app/src/main/java/com/example/test/MainActivity.kt
package com.example.test

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import android.Manifest
import android.content.pm.PackageManager
import com.example.test.util.AudioRecorderHelper
import com.example.test.widget.VoiceWaveView
import kotlin.math.sin

class MainActivity : AppCompatActivity() {

    private lateinit var voiceWaveView: VoiceWaveView
    private val audioHelper = AudioRecorderHelper()
    
    // 用于定时刷新的 Handler
    private val handler = Handler(Looper.getMainLooper())
    private val REQUEST_AUDIO_PERMISSION_CODE = 200
    
    // 测试模式开关
    private var isTestMode = true // [开关] 设为 true 开启正弦波测试，false 开启录音
    private var sineWaveTime = 0.0

    // 注意：我们将刷新率从 100ms 提高到 20ms，提高数据密度
    // 这不会加快波形滚动的速度（滚动速度由 View 内部的 speedPxPerMs 决定）
    // 这只会让波形的形状更细致
    private val DATA_REFRESH_INTERVAL = 20L

    // 刷新任务：读取音量 -> 更新 UI -> 再次调度
    private val updateTask = object : Runnable {
        override fun run() {
            if (isTestMode) {
                // 每次增加的角度变小，因为刷新变快了，保证波形频率正常
                sineWaveTime += 0.05
                // sin 的结果是 -1~1，我们转换到 0~1
                val amplitude = (sin(sineWaveTime) + 1) / 2.0
                voiceWaveView.addAmplitude(amplitude.toFloat())
            } else {
                // --- 场景 B：真实录音 ---
                val maxAmplitude = audioHelper.getCurrentAmplitude()
                val ratio = maxAmplitude / 32767f * 1.5f
                voiceWaveView.addAmplitude(ratio)
            }

            // 保持生产数据的频率与动画消耗的频率一致
            handler.postDelayed(this, DATA_REFRESH_INTERVAL)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        voiceWaveView = findViewById(R.id.voiceWaveView)

        // 检查并申请权限
        if (checkPermissions()) {
            startAudioVisualizer()
        } else {
            requestPermissions()
        }
    }

    private fun startAudioVisualizer() {
        if (!isTestMode) {
            audioHelper.startRecord(cacheDir)
        }
        handler.post(updateTask)
    }

    private fun stopAudioVisualizer() {
        // 停止 UI 刷新
        handler.removeCallbacks(updateTask)
        // 停止录音
        if (!isTestMode) {
            audioHelper.stopRecord()
        }
        // 清空画布 (可选)
        // voiceWaveView.clear()
    }

    // --- 生命周期管理 (关键) ---
    override fun onResume() {
        super.onResume()
        if (checkPermissions()) {
            startAudioVisualizer()
        }
    }

    override fun onPause() {
        super.onPause()
        stopAudioVisualizer() // 切后台时必须停止，否则会在后台占用麦克风
    }

    // --- 权限相关样板代码 ---
    private fun checkPermissions(): Boolean {
        return if (isTestMode) {
            true // 测试模式不需要权限
        } else {
            val result = ContextCompat.checkSelfPermission(applicationContext, Manifest.permission.RECORD_AUDIO)
            result == PackageManager.PERMISSION_GRANTED
        }
    }

    private fun requestPermissions() {
        if (!isTestMode) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                REQUEST_AUDIO_PERMISSION_CODE
            )
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_AUDIO_PERMISSION_CODE) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startAudioVisualizer()
            } else {
                Toast.makeText(this, "需要录音权限才能展示波形", Toast.LENGTH_LONG).show()
            }
        }
    }
}